cpp/src/arrow/acero/concurrent_queue_internal.h

```diff
@@ -65,17 +65,19 @@ class ConcurrentQueue {
     return queue_.empty();
   }
 
-  // Un-synchronized access to front
-  // For this to be "safe":
-  // 1) the caller logically guarantees that queue is not empty
-  // 2) pop/try_pop cannot be called concurrently with this
-  const T& UnsyncFront() const { return queue_.front(); }
-
-  size_t UnsyncSize() const { return queue_.size(); }
+  const T& Front() const {
+    // Need to lock the queue because `front()` may be implemented in terms
+    // of `begin()`, which isn't safe with concurrent calls to e.g. `push()`.
```

Comment:
is this `push` means `deque::push_back/front` or `ConcurrentQueue::Push()`?

Reply:
Here it means `std::queue::push` :) (which is concretely the same as `std::deque::push_back`)

Reply:
Ah, the original code is too optimistic. The deque might implemented as something like `std::vector<block*>` and `block*` for elements, the `front()` might visit `std::vector<block*>`  and `push_back` might enlarge it, so it really has problem here